src/main/java/com/google/cloud/genomics/dataflow/readers/bam/ReadBAMTransform.java

```diff
@@ -0,0 +1,142 @@
+import com.google.cloud.genomics.utils.Contig;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Takes a tuple of 2 collections: Contigs and BAM files and transforms them into
+ * a collection of reads by reading BAM files in a sharded manner.
+ */
+public class ReadBAMTransform extends PTransform<PCollectionTuple, PCollection<Read>> {
+  GenomicsFactory.OfflineAuth auth;
+  
+  public static TupleTag<Contig> CONTIGS_TAG = new TupleTag<>();
+  public static TupleTag<String> BAMS_TAG = new TupleTag<>();
+  
+  public static PCollection<Read> getReadsFromBAMFilesSharded(
+      Pipeline p, 
+      GenomicsFactory.OfflineAuth auth,
+      Iterable<Contig> contigs, List<String> BAMFiles) {
+      ReadBAMTransform readBAMSTransform = new ReadBAMTransform();
+      readBAMSTransform.setAuth(auth);
+      PCollectionTuple tuple = PCollectionTuple
+          .of(
+              ReadBAMTransform.BAMS_TAG, 
+                p.apply(
+                    Create.of(BAMFiles))
+                .setCoder(StringUtf8Coder.of()))
+         .and(ReadBAMTransform.CONTIGS_TAG, 
+             p.apply(
+                 Create.of(
+                     contigs))
+             .setCoder(SerializableCoder.of(Contig.class)));
+      return readBAMSTransform.apply(tuple);
+  }
+  
+  public static class ShardFn extends DoFn<String, BAMShard> {
+    GenomicsFactory.OfflineAuth auth;
+    PCollectionView<Iterable<Contig>, ?> contigsView;
+    Storage.Objects storage;
+    
+    public ShardFn(GenomicsFactory.OfflineAuth auth, PCollectionView<Iterable<Contig>, ?> contigsView) {
+      this.auth = auth;
+      this.contigsView = contigsView;
+    }
+    
+    @Override
+    public void startBundle(DoFn<String, BAMShard>.Context c) throws IOException {
+      storage = GCSOptions.Methods.createStorageClient(c, auth);
+    }
+    
+    @Override
+    public void processElement(ProcessContext c) throws java.lang.Exception {
+      (new Sharder(storage, c.element(), c.sideInput(contigsView), c))
+          .process();
+    }
+  }
+  
+  public static class ReadFn extends DoFn<BAMShard, Read> {
+    GenomicsFactory.OfflineAuth auth;
+    Storage.Objects storage;
+    
+    public ReadFn(GenomicsFactory.OfflineAuth auth) {
+      this.auth = auth;
+
+    }
+    
+    @Override
+    public void startBundle(DoFn<BAMShard, Read>.Context c) throws IOException {
+      storage = GCSOptions.Methods.createStorageClient(c, auth);
+    }
+    
+    @Override
+    public void processElement(ProcessContext c) throws java.lang.Exception {
+      (new Reader(storage, c.element(), c))
+          .process();
+    }
+  }
+  
+  
+  @Override
+  public PCollection<Read> apply(PCollectionTuple contigsAndBAMs) {
+
+    final PCollection<Contig> contigs = contigsAndBAMs.get(CONTIGS_TAG);
+    final PCollectionView<Iterable<Contig>, ?> contigsView =
```

Comment:
Hmm, I disagree: we want to make it clear that to apply a ReadBAMTransform you need to pass it both files and contigs. The interface makes it explicit.
If we were to use a side input then clients would have to call the transform on the PCollection of e.g. files and supply contigs as side input or vice versa and the call would look weird.
Now, in the guts of implementation we DO need to make a choice because ParDo forces the issue and makes us chose the main collection to iterate on. But I could see us implement a logic where if size of contigs is large and BAMs are few, we actually do ParDo over contigs and pass BAMs as side input. Its a bit contrived but serves to illustrate that we would want to hide decision making like this from the callers of the transform. To them, they should be passing 2 equally important collections: BAMs and Contigs. I also really hate how sideInput is implemented: you have to pass it twice - once to have an instance variable (e.g. in the constructor) and once in the explicit call to .withSideInput. Suppose you do first and not the second - you get ugly ugly errors. So I want to avoid it if possible.